content/blog/2022/2022-02-15-getting-started-with-spring-webflux.md

```diff
@@ -0,0 +1,566 @@
+                        .by(Collections.singletonList(Sort.Order.asc("age")))
+                );
+        query.addCriteria(Criteria
+                .where("name")
+                .regex(name)
+        );
+
+        return reactiveMongoTemplate
+                .find(query, User.class);
+    }
+}
+```
+
+We have defined service methods to save, update, fetch, search and delete a user. We have primarily used *UserRepository* to store and retrieve data from MongoDB, but we have also used a *ReactiveTemplate* and *Query* to search for a user given by a regex string.
+
+#### Web Layer
+
+We have covered the middleware layers to store and retrieve data, let’s just focus on the Web layer. Spring Webflux basically supports two programming models:
+
+* Annotation-based Reactive components
+* Functional Routing and Handling
+
+##### Annotation-based Reactive Components
+
+Let’s first look into the Annotation based components. We can simply create a `UserController` for that and annotate with routes and methods.
+
+```java
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/users")
+public class UserController {
+
+    private final UserService userService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public Mono<User> create(@RequestBody User user){
+        return userService.createUser(user);
+    }
+
+    @GetMapping
+    public Flux<User> getAllUsers(){
+        return userService.getAllUsers();
+    }
+
+    @GetMapping("/{userId}")
+    public Mono<ResponseEntity<User>> getUserById(@PathVariable String userId){
+        Mono<User> user = userService.findById(userId);
+        return user.map(ResponseEntity::ok)
+                .defaultIfEmpty(ResponseEntity.notFound().build());
+    }
+
+    @PutMapping("/{userId}")
+    public Mono<ResponseEntity<User>> updateUserById(@PathVariable String userId, @RequestBody User user){
+        return userService.updateUser(userId,user)
+                .map(ResponseEntity::ok)
+                .defaultIfEmpty(ResponseEntity.badRequest().build());
+    }
+
+    @DeleteMapping("/{userId}")
+    public Mono<ResponseEntity<Void>> deleteUserById(@PathVariable String userId){
+        return userService.deleteUser(userId)
+                .map( r -> ResponseEntity.ok().<Void>build())
+                .defaultIfEmpty(ResponseEntity.notFound().build());
+    }
+
+    @GetMapping("/search")
+    public Flux<User> searchUsers(@RequestParam("name") String name) {
+        return userService.fetchUsers(name);
+    }
+}
+```
+
+This almost looks same as the controller defined in Spring MVC. But the major difference between Spring MVC and Spring Webflux relies on how the request and response are handled using non-blocking Publishers, *Mono* and *Flux*. We don’t need to call subscribe methods in the Controller as the internal classes of Spring would call it for us at the right time.
+
+> *Note: We must make sure that we don’t use any blocking methods throughout the lifecycle of an API.*
```

Comment:
This is better highlighted as a "warning" box. See here: https://reflectoring.io/contribute/writing-guide/#asides

Reply:
Added a warning box.